plugins/hls-fourmolu-plugin/src/Ide/Plugin/Fourmolu.hs

```diff
@@ -2,91 +2,127 @@
+{-# LANGUAGE DataKinds                #-}
+{-# LANGUAGE OverloadedLabels         #-}
 
 module Ide.Plugin.Fourmolu (
     descriptor,
     provider,
 ) where
 
-import           Control.Exception               (try)
+import           Control.Exception               (IOException, try)
 import           Control.Lens                    ((^.))
+import           Control.Monad
 import           Control.Monad.IO.Class
 import           Data.Bifunctor                  (first)
+import           Data.Maybe
 import qualified Data.Text                       as T
+import qualified Data.Text.IO                    as T
 import           Development.IDE                 hiding (pluginHandlers)
 import           Development.IDE.GHC.Compat      as Compat hiding (Cpp)
 import qualified Development.IDE.GHC.Compat.Util as S
 import           GHC.LanguageExtensions.Type     (Extension (Cpp))
-import           Ide.PluginUtils                 (makeDiffTextEdit)
+import           Ide.Plugin.Properties
+import           Ide.PluginUtils                 (makeDiffTextEdit, usePropertyLsp)
 import           Ide.Types
 import           Language.LSP.Server             hiding (defaultConfig)
 import           Language.LSP.Types
 import           Language.LSP.Types.Lens         (HasTabSize (tabSize))
 import           Ormolu
+import           System.Exit
 import           System.FilePath
-
--- ---------------------------------------------------------------------
+import           System.IO                       (stderr)
+import           System.Process.Run              (proc, cwd)
+import           System.Process.Text             (readCreateProcessWithExitCode)
 
 descriptor :: PluginId -> PluginDescriptor IdeState
 descriptor plId =
     (defaultPluginDescriptor plId)
-        { pluginHandlers = mkFormattingHandlers provider
+        { pluginHandlers = mkFormattingHandlers $ provider plId
         }
 
--- ---------------------------------------------------------------------
-
-provider :: FormattingHandler IdeState
-provider ideState typ contents fp fo = withIndefiniteProgress title Cancellable $ do
-    ghc <- liftIO $ runAction "Fourmolu" ideState $ use GhcSession fp
-    fileOpts <- case hsc_dflags . hscEnv <$> ghc of
-        Nothing -> return []
-        Just df -> liftIO $ convertDynFlags df
-
-    let format printerOpts =
-            first (responseError . ("Fourmolu: " <>) . T.pack . show)
-                <$> try @OrmoluException (makeDiffTextEdit contents <$> ormolu config fp' (T.unpack contents))
-          where
-            config =
-                defaultConfig
-                    { cfgDynOptions = fileOpts
-                    , cfgRegion = region
-                    , cfgDebug = True
-                    , cfgPrinterOpts =
-                        fillMissingPrinterOpts
-                            (printerOpts <> lspPrinterOpts)
-                            defaultPrinterOpts
-                    }
+properties :: Properties '[ 'PropertyKey "cli" 'TBoolean]
```

Comment:
bikeshedding: "external"? "shell"? I think "cli" seems fine though.

Reply:
Yeah, surpisingly no word I could think of here seems perfect

Reply:
I've decided I do prefer "external": [fb3975e](https://github.com/haskell/haskell-language-server/pull/2763/commits/fb3975eedbf553457d46c74ea1500182aa3320b3). Not going to bother changing the variable names though.